buffer/src/main/java/io/netty/buffer/SimpleLeakAwareByteBuf.java

```diff
@@ -16,64 +16,93 @@
+     * is called this object will be used as the argument. It is also assumed that this object is used when
+     * {@link ResourceLeakDetector#track(Object)} is called to create {@link #leak}.
+     */
+    private final ByteBuf trackedByteBuf;
+    final ResourceLeakTracker<ByteBuf> leak;
 
-    SimpleLeakAwareByteBuf(ByteBuf buf, ResourceLeak leak) {
-        super(buf);
-        this.leak = leak;
+    SimpleLeakAwareByteBuf(ByteBuf wrapped, ByteBuf trackedByteBuf, ResourceLeakTracker<ByteBuf> leak) {
+        super(wrapped);
+        this.trackedByteBuf = ObjectUtil.checkNotNull(trackedByteBuf, "trackedByteBuf");
+        this.leak = ObjectUtil.checkNotNull(leak, "leak");
+    }
+
+    SimpleLeakAwareByteBuf(ByteBuf wrapped, ResourceLeakTracker<ByteBuf> leak) {
+        this(wrapped, wrapped, leak);
     }
 
     @Override
-    public boolean release() {
-        boolean deallocated =  super.release();
-        if (deallocated) {
-            leak.close();
-        }
-        return deallocated;
+    public ByteBuf slice() {
+        return newSharedLeakAwareByteBuf(super.slice());
     }
 
     @Override
-    public boolean release(int decrement) {
-        boolean deallocated = super.release(decrement);
-        if (deallocated) {
-            leak.close();
-        }
-        return deallocated;
+    public ByteBuf slice(int index, int length) {
+        return newSharedLeakAwareByteBuf(super.slice(index, length));
     }
 
     @Override
-    public ByteBuf order(ByteOrder endianness) {
-        leak.record();
-        if (order() == endianness) {
-            return this;
-        } else {
-            return new SimpleLeakAwareByteBuf(super.order(endianness), leak);
-        }
+    public ByteBuf duplicate() {
+        return newSharedLeakAwareByteBuf(super.duplicate());
     }
 
     @Override
-    public ByteBuf slice() {
-        return new SimpleLeakAwareByteBuf(super.slice(), leak);
+    public ByteBuf readSlice(int length) {
+        return newSharedLeakAwareByteBuf(super.readSlice(length));
     }
 
     @Override
-    public ByteBuf slice(int index, int length) {
-        return new SimpleLeakAwareByteBuf(super.slice(index, length), leak);
+    public final boolean release() {
+        if (super.release()) {
+            closeLeak();
+            return true;
+        }
+        return false;
     }
 
     @Override
-    public ByteBuf duplicate() {
-        return new SimpleLeakAwareByteBuf(super.duplicate(), leak);
+    public final boolean release(int decrement) {
+        if (super.release(decrement)) {
+            closeLeak();
+            return true;
+        }
+        return false;
+    }
+
+    private void closeLeak() {
+        // Close the ResourceLeakTracker with the tracked ByteBuf as argument. This must be the same that was used when
+        // calling DefaultResourceLeak.track(...).
+        boolean closed = leak.close(trackedByteBuf);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "closed". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)